### PR TITLE
[SC-890] Feature/get rid of truffle

### PR DESCRIPTION
**Pull request by SevenSwen**

### CODE REVIEW

File: hardhat.config.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
+require('@nomiclabs/hardhat-etherscan');
```
Comment (ZumZoom) [inline hardhat.config.js:1]: ?

File: package.json [hunk 2]
```diff
@@ -8,37 +8,35 @@
...
-    "dotenv": "16.0.1",
-    "eslint": "8.16.0",
-    "eslint-config-standard": "17.0.0",
-    "eslint-plugin-import": "2.26.0",
-    "eslint-plugin-n": "15.2.0",
-    "eslint-plugin-promise": "6.0.0",
+    "@metamask/eth-sig-util": "6.0.0",
+    "@nomicfoundation/hardhat-chai-matchers": "2.0.2",
+    "@nomicfoundation/hardhat-ethers": "3.0.4",
+    "@nomiclabs/hardhat-etherscan": "3.1.7",
```
Comment (ZumZoom) [inline package.json:18]: ?

File: contracts/CumulativeMerkleDrop.sol [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
 
-import "./interfaces/ICumulativeMerkleDrop.sol";
+import { ICumulativeMerkleDrop } from "./interfaces/ICumulativeMerkleDrop.sol";
 
 contract CumulativeMerkleDrop is Ownable, ICumulativeMerkleDrop {
     using SafeERC20 for IERC20;
     // using MerkleProof for bytes32[];
+    error InvalidProof();
+    error NothingToClaim();
+    error MerkleRootWasUpdated();
```
Comment (ZumZoom) [inline contracts/CumulativeMerkleDrop.sol:16]: move errors to interface

File: contracts/CumulativeMerkleDrop128.sol [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
+import { SafeERC20, IERC20 } from  "@1inch/solidity-utils/contracts/libraries/SafeERC20.sol";
 
-import "./interfaces/ICumulativeMerkleDrop128.sol";
+import { ICumulativeMerkleDrop128 } from  "./interfaces/ICumulativeMerkleDrop128.sol";
 
 contract CumulativeMerkleDrop128 is Ownable, ICumulativeMerkleDrop128 {
     using SafeERC20 for IERC20;
+    error InvalidProof();
+    error NothingToClaim();
+    error MerkleRootWasUpdated();
```
Comment (ZumZoom) [inline contracts/CumulativeMerkleDrop128.sol:14]: move errors to interface

File: contracts/SignatureMerkleDrop128.sol [hunk 5]
```diff
@@ -1,23 +1,26 @@
...
+import { ECDSA } from  "@1inch/solidity-utils/contracts/libraries/ECDSA.sol";
 
-import "./interfaces/ISignatureMerkleDrop128.sol";
+import { ISignatureMerkleDrop128 } from  "./interfaces/ISignatureMerkleDrop128.sol";
 
 contract SignatureMerkleDrop128 is ISignatureMerkleDrop128, Ownable {
     using Address for address payable;
     using SafeERC20 for IERC20;
+    error InvalidProof();
+    error DropAlreadyClaimed();
```
Comment (ZumZoom) [inline contracts/SignatureMerkleDrop128.sol:17]: move errors to interface

File: test/CumulativeMerkleDrop.test.js [hunk 6]
```diff
@@ -11,84 +14,98 @@ const {
...
-            await this.drop.contract.methods.verify(this.proofs[findSortedIndex(this, 0)], this.root, this.leaves[0]).send({ from: _ });
-            expect(await this.drop.verify(this.proofs[findSortedIndex(this, 0)], this.root, this.leaves[0])).to.be.true;
-        }
-        // await this.drop.contract.methods.verifyAsm(this.proofs[findSortedIndex(this, 0)], this.root, this.leaves[0]).send({ from: _ });
-        // expect(await this.drop.verifyAsm(this.proofs[findSortedIndex(this, 0)], this.root, this.leaves[0])).to.be.true;
-        await this.drop.claim(accounts[0], 1, this.root, this.proofs[findSortedIndex(this, 0)]);
-    });
+        return {
+            accounts: { owner, alice, bob, carol, dan },
+            constracts: { token, drop },
```
Comment (ZumZoom) [inline test/CumulativeMerkleDrop.test.js:50]:
```suggestion
            contracts: { token, drop },
```

File: test/CumulativeMerkleDrop.test.js [hunk 7]
```diff
@@ -11,84 +14,98 @@ const {
...
-    describe('Single drop for 4 wallets: [1, 2, 3, 4]', async function () {
-        beforeEach(async function () {
-            const { hashedElements, leaves, root, proofs } = await makeDrop(this.token, this.drop, [w1, w2, w3, w4], [1, 2, 3, 4], 10);
-            this.hashedElements = hashedElements;
-            this.leaves = leaves;
-            this.root = root;
-            this.proofs = proofs;
-        });
+    it('Benchmark 30000 wallets (merkle tree height 15)', async function () {
+        const { accounts: { alice }, constracts: { token, drop } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/CumulativeMerkleDrop.test.js:55]:
```suggestion
        const { accounts: { alice }, contracts: { token, drop } } = await loadFixture(deployContractsFixture);
```

File: test/CumulativeMerkleDrop128.test.js [hunk 8]
```diff
@@ -10,93 +13,106 @@ const {
...
-            this.proofs = proofs;
-            this.salts = salts;
-        });
+        const { token, drop } = await initContracts();
+        await Promise.all([alice, bob, carol, dan].map(w => token.mint(w, 1n)));
 
-        shouldBehaveLikeMerkleDropFor4WalletsWithBalances1234('CMD', [w1, w2, w3, w4], findSortedIndex, true);
+        return {
+            accounts: { owner, alice, bob, carol, dan },
+            constracts: { token, drop },
```
Comment (ZumZoom) [inline test/CumulativeMerkleDrop128.test.js:56]:
```suggestion
            contracts: { token, drop },
```

File: test/CumulativeMerkleDrop128.test.js [hunk 9]
```diff
@@ -10,93 +13,106 @@ const {
...
+        };
+    }
+
+    it('Benchmark 30000 wallets (merkle tree height 15)', async function () {
+        const { accounts: { alice }, constracts: { token, drop } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/CumulativeMerkleDrop128.test.js:61]:
```suggestion
        const { accounts: { alice }, contracts: { token, drop } } = await loadFixture(deployContractsFixture);
```

File: test/behaviors/MerkleDrop.behavior.js [hunk 10]
```diff
@@ -1,133 +1,62 @@
...
-                        is128version
-                            ? await this.drop.claim(this.salts[3], wallets[3], 4, this.root, this.proofs[findSortedIndex(this, 3)])
-                            : await this.drop.claim(wallets[3], 4, this.root, this.proofs[findSortedIndex(this, 3)])
-                    ),
-                    'Claimed', claimedEvent(wallets[3], '4'),
-                );
-            });
+                it('should fail to claim second time', async function () {
+                    const {
+                        constracts: { drop },
```
Comment (ZumZoom) [inline test/behaviors/MerkleDrop.behavior.js:42]:
```suggestion
                        contracts: { drop },
```

File: test/behaviors/MerkleDrop.behavior.js [hunk 11]
```diff
@@ -1,133 +1,62 @@
...
-                if (is128version) {
-                    await this.drop.claim(this.salts[2], wallets[2], 3, this.root, this.proofs[findSortedIndex(this, 2)]);
-                } else {
-                    await this.drop.claim(wallets[2], 3, this.root, this.proofs[findSortedIndex(this, 2)]);
-                }
+        for (let i = 0; i < walletsCount; i++) {
+            describe(`Wallet ${i + 1}`, function () {
+                it('should succeed to claim', async function () {
+                    const {
+                        constracts: { drop },
```
Comment (ZumZoom) [inline test/behaviors/MerkleDrop.behavior.js:29]:
```suggestion
                        contracts: { drop },
```

File: test/behaviors/MerkleDrop.behavior.js [hunk 12]
```diff
@@ -1,133 +1,62 @@
...
-                    (
-                        is128version
-                            ? await this.drop.claim(this.salts[2], wallets[2], 3, this.root, this.proofs[findSortedIndex(this, 2)])
-                            : await this.drop.claim(wallets[2], 3, this.root, this.proofs[findSortedIndex(this, 2)])
-                    ),
-                    'Claimed', claimedEvent(wallets[2], '3'),
-                );
-            });
+            return {
+                constracts: { drop },
```
Comment (ZumZoom) [inline test/behaviors/MerkleDrop.behavior.js:20]:
```suggestion
                contracts: { drop },
```

File: test/behaviors/CumulativeMerkleDrop.behavior.js [hunk 13]
```diff
@@ -1,110 +1,75 @@
...
-            await expectRevert(
-                (
-                    is128version
-                        ? this.drop.claim(this.salts[0], wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                        : this.drop.claim(wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                ),
-                `${errorPrefix}: Nothing to claim`,
-            );
+            const {
+                constracts: { token, drop },
```
Comment (ZumZoom) [inline test/behaviors/CumulativeMerkleDrop.behavior.js:56]:
```suggestion
                contracts: { token, drop },
```

File: test/SignatureMerkleDrop128.test.js [hunk 14]
```diff
@@ -1,114 +1,102 @@
...
-            this.hashedElements = hashedElements;
-            this.leaves = leaves;
-            this.root = root;
-            this.proofs = proofs;
-            this.drop = drop;
-            this.account = Wallet.fromPrivateKey(Buffer.from('ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80', 'hex'));
-        });
+        return {
+            accounts: { owner, alice, bob, carol, dan },
+            constracts: { token, drop },
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:62]:
```suggestion
            contracts: { token, drop },
```

File: test/SignatureMerkleDrop128.test.js [hunk 15]
```diff
@@ -1,114 +1,102 @@
...
 
-        it('Should enumerate items properly', async function () {
-            for (let i = 0; i < this.proofs.length; i++) {
-                const result = await this.drop.verify(this.proofs[findSortedIndex(this, i)], this.root, this.leaves[findSortedIndex(this, i)]);
-                expect(result.valid).to.be.true;
-                expect(BigInt(result.index)).to.be.equal(BigInt(findSortedIndex(this, i)));
-            }
-        });
+    it('Should enumerate items properly', async function () {
+        const { constracts: { drop }, others: { hashedElements, leaves, proofs } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:68]:
```suggestion
        const { contracts: { drop }, others: { hashedElements, leaves, proofs } } = await loadFixture(deployContractsFixture);
```

File: test/SignatureMerkleDrop128.test.js [hunk 16]
```diff
@@ -1,114 +1,102 @@
...
+        }
+    });
 
-        it('Should transfer money to another wallet', async function () {
-            const data = MerkleTree.bufferToHex(keccak256(w1));
-            const signature = personalSign({ privateKey: this.account.getPrivateKey(), data });
-            await this.drop.claim(w1, 1, this.proofs[findSortedIndex(this, 0)], signature);
-        });
+    it('Should transfer money to another wallet', async function () {
+        const { accounts: { alice }, constracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:78]:
```suggestion
        const { accounts: { alice }, contracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```

File: test/SignatureMerkleDrop128.test.js [hunk 17]
```diff
@@ -1,114 +1,102 @@
...
 
-        it('Should disallow invalid proof', async function () {
-            const data = MerkleTree.bufferToHex(keccak256(w1));
-            const signature = personalSign({ privateKey: this.account.getPrivateKey(), data });
-            await expectRevert(
-                this.drop.claim(w1, 1, '0x', signature),
-                'MD: Invalid proof');
-        });
+    it('Should disallow invalid proof', async function () {
+        const { accounts: { alice }, constracts: { drop }, others: { signature } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:83]:
```suggestion
        const { accounts: { alice }, contracts: { drop }, others: { signature } } = await loadFixture(deployContractsFixture);
```

File: test/SignatureMerkleDrop128.test.js [hunk 18]
```diff
@@ -1,114 +1,102 @@
...
 
-        it('Should disallow invalid receiver', async function () {
-            const data = MerkleTree.bufferToHex(keccak256(w1));
-            const signature = personalSign({ privateKey: this.account.getPrivateKey(), data });
-            await expectRevert(
-                this.drop.claim(w2, 1, this.proofs[findSortedIndex(this, 0)], signature),
-                'MD: Invalid proof');
-        });
+    it('Should disallow invalid receiver', async function () {
+        const { accounts: { bob }, constracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:90]:
```suggestion
        const { accounts: { bob }, contracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```

File: test/SignatureMerkleDrop128.test.js [hunk 19]
```diff
@@ -1,114 +1,102 @@
...
 
-        it('Should disallow double claim', async function () {
-            const data = MerkleTree.bufferToHex(keccak256(w1));
-            const signature = personalSign({ privateKey: this.account.getPrivateKey(), data });
-            const fn = () => this.drop.claim(w1, 1, this.proofs[findSortedIndex(this, 0)], signature);
-            await fn();
-            await expectRevert(fn(), 'MD: Drop already claimed');
-        });
+    it('Should disallow double claim', async function () {
+        const { accounts: { alice }, constracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```
Comment (ZumZoom) [inline test/SignatureMerkleDrop128.test.js:97]:
```suggestion
        const { accounts: { alice }, contracts: { drop }, others: { hashedElements, leaves, proofs, signature } } = await loadFixture(deployContractsFixture);
```

File: test/behaviors/CumulativeMerkleDrop.behavior.js [hunk 20]
```diff
@@ -1,110 +1,75 @@
...
-                (
-                    is128version
-                        ? this.drop.claim(this.salts[0], wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                        : this.drop.claim(wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                ),
-                `${errorPrefix}: Nothing to claim`,
-            );
-        });
+            return {
+                constracts: { token, drop },
```
Comment (ZumZoom) [inline test/behaviors/CumulativeMerkleDrop.behavior.js:20]:
```suggestion
                contracts: { token, drop },
```

File: test/behaviors/CumulativeMerkleDrop.behavior.js [hunk 21]
```diff
@@ -1,110 +1,75 @@
...
-            await expectEvent(
-                (
-                    is128version
-                        ? await this.drop.claim(this.salts[0], wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                        : await this.drop.claim(wallets[0], 3, this.root, this.proofs[findSortedIndex(this, 0)])
-                ),
-                'Claimed', claimedEvent(wallets[0], '3'),
-            );
+            const {
+                constracts: { token, drop },
```
Comment (ZumZoom) [inline test/behaviors/CumulativeMerkleDrop.behavior.js:28]:
```suggestion
                contracts: { token, drop },
```

**Review by ZumZoom — CHANGES_REQUESTED** [review]

File: contracts/SignatureMerkleDrop128.sol (hunk 5 above)
Comment (SevenSwen) [reply contracts/SignatureMerkleDrop128.sol:17]: fixed

File: contracts/CumulativeMerkleDrop.sol (hunk 3 above)
Comment (SevenSwen) [reply contracts/CumulativeMerkleDrop.sol:16]: fixed

File: contracts/CumulativeMerkleDrop128.sol (hunk 4 above)
Comment (SevenSwen) [reply contracts/CumulativeMerkleDrop128.sol:14]: fixed

File: test/SignatureMerkleDrop128.test.js (hunk 17 above)
Comment (SevenSwen) [reply test/SignatureMerkleDrop128.test.js]: fixed

File: test/SignatureMerkleDrop128.test.js (hunk 19 above)
Comment (SevenSwen) [reply test/SignatureMerkleDrop128.test.js]: fixed

File: test/behaviors/CumulativeMerkleDrop.behavior.js (hunk 20 above)
Comment (SevenSwen) [reply test/behaviors/CumulativeMerkleDrop.behavior.js]: fixed

File: test/behaviors/CumulativeMerkleDrop.behavior.js (hunk 21 above)
Comment (SevenSwen) [reply test/behaviors/CumulativeMerkleDrop.behavior.js]: fixed

File: test/SignatureMerkleDrop128.test.js (hunk 18 above)
Comment (SevenSwen) [reply test/SignatureMerkleDrop128.test.js]: fixed

File: hardhat.config.js (hunk 1 above)
Comment (SevenSwen) [reply hardhat.config.js]: fixed

File: package.json (hunk 2 above)
Comment (SevenSwen) [reply package.json]: fixed

**Pull request merged**